File: modules/agent/src/services/provenance_service.py
"""Service layer for data provenance business logic."""

import logging
from typing import Optional, Any, List

from fastapi import HTTPException

from repositories.data_provenance_repository import (
    find_provenance,
    create_provenance as create_provenance_repo,
    ProvenanceCreate,
)

# Re-export Pydantic models for controllers
__all__ = ["ProvenanceCreate"]

logger = logging.getLogger(__name__)

VALID_ENTITY_TYPES = ("Organization", "Individual")


async def get_provenance(entity_type: str, entity_id: int, field_name: Optional[str] = None) -> List:
    """Get provenance records for an entity."""
    if entity_type not in VALID_ENTITY_TYPES:
        raise HTTPException(status_code=400, detail="Invalid entity_type")
    return await find_provenance(entity_type, entity_id, field_name)


async def create_provenance(
    entity_type: str,
    entity_id: int,
    field_name: str,
    source: str,
    user_id: Optional[int],
    source_url: Optional[str] = None,
    confidence: Optional[float] = None,
    raw_value: Optional[Any] = None,
):
    """Create a new provenance record."""
    if entity_type not in VALID_ENTITY_TYPES:
        raise HTTPException(status_code=400, detail="Invalid entity_type")

    return await create_provenance_repo(
        entity_type=entity_type,
        entity_id=entity_id,
        field_name=field_name,
        source=source,
        source_url=source_url,
        confidence=confidence,
        verified_by=user_id,
        raw_value=raw_value,
    )

File: modules/agent/internal/services/provenance_service.go
package services

import (
	"errors"

	"agent/internal/repositories"
	"agent/internal/schemas"
	"agent/internal/serializers"
)

var ErrInvalidEntityType = errors.New("invalid entity_type: must be Organization or Individual")

// ProvenanceService handles provenance business logic
type ProvenanceService struct {
	provenanceRepo *repositories.ProvenanceRepository
}

// NewProvenanceService creates a new provenance service
func NewProvenanceService(provenanceRepo *repositories.ProvenanceRepository) *ProvenanceService {
	return &ProvenanceService{provenanceRepo: provenanceRepo}
}

func (s *ProvenanceService) isValidEntityType(entityType string) bool {
	return entityType == "Organization" || entityType == "Individual"
}

// GetProvenance returns provenance records for an entity
func (s *ProvenanceService) GetProvenance(entityType string, entityID int64, fieldName *string) ([]serializers.ProvenanceResponse, error) {
	if !s.isValidEntityType(entityType) {
		return nil, ErrInvalidEntityType
	}
	records, err := s.provenanceRepo.FindByEntity(entityType, entityID, fieldName)
	if err != nil {
		return nil, err
	}
	return serializers.ProvenancesToResponse(records), nil
}

// CreateProvenance creates a new provenance record
func (s *ProvenanceService) CreateProvenance(input schemas.ProvenanceCreate, userID *int64) (*serializers.ProvenanceResponse, error) {
	if !s.isValidEntityType(input.EntityType) {
		return nil, ErrInvalidEntityType
	}

	repoInput := repositories.ProvenanceCreateInput{
		EntityType: input.EntityType,
		EntityID:   input.EntityID,
		FieldName:  input.FieldName,
		Source:     input.Source,
		SourceURL:  input.SourceURL,
		Confidence: input.Confidence,
		VerifiedBy: userID,
		RawValue:   input.RawValue,
	}

	record, err := s.provenanceRepo.Create(repoInput)
	if err != nil {
		return nil, err
	}
	resp := serializers.ProvenanceToResponse(record)
	return &resp, nil
}
